lucene/core/src/java/org/apache/lucene/search/LRUQueryCache.java

```diff
@@ -832,5 +931,43 @@ public BulkScorer bulkScorer(LeafReaderContext context) throws IOException {
       return new DefaultBulkScorer(new ConstantScoreScorer(this, 0f, ScoreMode.COMPLETE_NO_SCORES, disi));
     }
 
+    // Perform a cache load asynchronously
+    // @return true if synchronous caching is needed, false otherwise
+    private boolean cacheAsynchronously(LeafReaderContext context, IndexReader.CacheHelper cacheHelper) {
+      /*
+       * If the current query is already being asynchronously cached,
+       * do not trigger another cache operation
+       */
+      if (inFlightAsyncLoadQueries.add(new MapKey(in.getQuery(),
+          cacheHelper.getKey())) == false) {
+        return false;
+      }
+
+      FutureTask<Void> task = new FutureTask<>(() -> {
+        try {
+          DocIdSet localDocIdSet = cache(context);
+          putIfAbsent(in.getQuery(), localDocIdSet, cacheHelper);
+        } finally {
+          // Remove the key from inflight
+          Object retValue = inFlightAsyncLoadQueries.remove(new MapKey(in.getQuery(), cacheHelper.getKey()));
```

Comment:
this call is protected by no lock while `inFlightAsyncLoadQueries` is not concurrent?

Reply:
Removed the map